ReactiveObjC/RACSignal+Operations.m

```diff
@@ -846,22 +846,57 @@ - (RACSignal *)takeUntilReplacement:(RACSignal *)replacement {
 
 - (RACSignal *)switchToLatest {
   return [[RACSignal createSignal:^(id<RACSubscriber> subscriber) {
-    RACMulticastConnection *connection = [self publish];
+    __block BOOL outerCompleted = NO;
+    __block BOOL innerCompleted = NO;
+    __block NSUInteger innerSignalIndex = 0;
 
-    RACDisposable *subscriptionDisposable = [[connection.signal
-      flattenMap:^(RACSignal *x) {
-        NSCAssert(x == nil || [x isKindOfClass:RACSignal.class], @"-switchToLatest requires that the source signal (%@) send signals. Instead we got: %@", self, x);
+    RACSerialDisposable *innerDisposable = [[RACSerialDisposable alloc] init];
+    NSLock *lock = [[NSLock alloc] init];
+
+    RACDisposable *outerDisposable = [self subscribeNext:^(RACSignal *signal) {
+      NSCAssert(signal == nil || [signal isKindOfClass:RACSignal.class], @"-switchToLatest expects signals as values, instead we got: %@", signal);
+      signal = signal ?: [RACSignal empty];
 
-        // -concat:[RACSignal never] prevents completion of the receiver from
-        // prematurely terminating the inner signal.
-        return [x takeUntil:[connection.signal concat:[RACSignal never]]];
-      }]
-      subscribe:subscriber];
+      [innerDisposable.disposable dispose];
+
+      [lock lock];
+      innerCompleted = NO;
+      ++innerSignalIndex;
+      NSUInteger signalIndex = innerSignalIndex;
+      [lock unlock];
```

Comment:
Might need to move this until after the `subscribeNext`.

Reply:
but then values that are sent synchronously will be sent within the lock. that sounds quite dangerous.

Reply:
I don't think that it's required. See my comment below on the order of things.

Reply:
@barakwei if you call `subscribeNext:` with the lock taken, and the call to `subscribeNext:` causes values to be sent synchronously, then the lock isn't released before values are sent.

I ran tests with your suggestion, and it seems that the RACCommandSpec gets stuck (probably deadlocked) on several tests.

Reply:
@barakwei if you call subscribeNext: with the lock taken, and the call to subscribeNext: causes values to be sent synchronously, then the lock isn't released before values are sent.

I ran tests with your suggestion, and it seems that the RACCommandSpec gets stuck (probably deadlocked) on several tests.

Reply:
My bad, I thought it was `createSignal:` and not `subscribeNext:`.